lib/http_utils.js

```diff
@@ -0,0 +1,244 @@
+ */
+function _handleHttpError(reqContext, callback, errInfos) {
+    const error = new Error('HTTP request failed');
+    error.infos = errInfos;
+    _updateOperationContext(reqContext, { error });
+    return callback(reqContext);
+}
+
+/**
+ * Create new HTTP request
+ *
+ * @param {Object} options to pass to the request
+ * @param {werelogs.Logger} logger to use
+ * @param {Object} reqContext Request context
+ * @param {function} callback to call on reply
+ * @return {http.ClientRequest} created request
+ */
+function _createHttpRequest(options, logger, reqContext, callback) {
+    const request = http.request(options, response => {
+        if (request.aborted) {
+            /* Request was already handled/stopped
+             * Abort caller should handle the error
+             * (calling callback and all) so we do nothing here.
+             */
+            return null;
+        }
+
+        request.done = true; // to avoid doing anything in 'timeout' handler
+
+        const success = Math.floor(response.statusCode / 200);
+        if (success !== 1) {
+            const errInfos = {
+                status: response.statusCode,
+                headers: response.headers,
+                method: options.method,
+            };
+            logger.error(`${response.method} ${response.url}:`, errInfos);
+            return _handleHttpError(reqContext, callback, errInfos);
+        }
+
+        return _handleHttpSuccess(reqContext, callback, response);
+    });
+
+    return request;
+}
+
+/**
+ * Send a HTTP request
+ *
+ * @param {Object} options Object filled with all necessary
+ *                 options and headers. See _getCommonStoreRequestOptions
+ * @param {werelogs.Logger} logger to use
+ * @param {Object} reqContext Request context
+ * @param {Number} timeoutMs Request timeout
+ * @param {function} callback (opContext) -> ?
+ * @return {http.ClientRequest} object
+ */
+function newRequest(options, logger, reqContext, timeoutMs, callback) {
+    const enhanceLogs = {
+        path: options.path,
+        port: options.port,
+        method: options.method,
+        host: options.hostname,
+    };
+
+    const request = _createHttpRequest(
+        options, logger, reqContext, callback);
+
+    // disable nagle algorithm
+    request.setNoDelay(true);
+
+    request.on('error', error => {
+        logger.error(error.message, enhanceLogs);
+        // TODO: avoid this hack?
+        /* eslint-disable no-param-reassign */
+        error.infos = enhanceLogs;
+        error.infos.status = 500; // HTTP return code
+        error.infos.method = options.method;
+        /* eslint-disable no-param-reassign */
+        _updateOperationContext(reqContext, { error });
+        const ret = callback(reqContext);
+        request.destroy();
+        return ret;
+    });
+
+    request.setTimeout(timeoutMs, () => {
+        /* Hack, since I had much trouble cancelling
+         * timeout event on success...
```

Comment:
a simple timeout.cancel() was not sufficient ? :o 
Maybe is it due to the event being fired already in the scheduler, that may not be cancellable anymore ?